### PR TITLE
Add missing `overlayColor` property in `styleFrom` methods

### LINKED ISSUE #146636: Add missing `overlayColor` property  in `styleFrom` methods
### Use case

A while ago https://github.com/flutter/flutter/pull/142762 added `overlayColor` in the `styleFrom` methods for the `TextButton`, `ElevatedButton`, `FilledButton`, and `OutlinedButton`. 
This made it easily to override color from `styleFrom` method.

However, there are other  `styleFrom` methods missing this prop.

Screenshot 2024-04-11 at 17 25 54

### Proposal

Add the missing missing `overlayColor` property in `styleFrom` methods

### PR DESCRIPTION
**Pull request by TahaTesser**
fixes [Add missing `overlayColor` property  in `styleFrom` methods](https://github.com/flutter/flutter/issues/146636)

### Code sample

<details>
<summary>expand to view the code sample</summary> 

```dart
import 'package:flutter/material.dart';

enum Sizes { extraSmall, small, medium, large, extraLarge }

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.spaceEvenly,
            children: <Widget>[
              Text('styleFrom(overlayColor: Colors.red)', style: Theme.of(context).textTheme.titleLarge),
              TextButton(
                style: TextButton.styleFrom(overlayColor: Colors.red),
                onPressed: () {},
                child: const Text('TextButton'),
              ),
              IconButton(
                style: IconButton.styleFrom(
                  overlayColor: Colors.red,
                ),
                onPressed: () {},
                icon: const Icon(Icons.add),
              ),
              MenuBar(
                children: [
                  MenuItemButton(
                    style: MenuItemButton.styleFrom(overlayColor: Colors.red),
                    child: const Text('MenuItemButton'),
                    onPressed: () {},
                  ),
                  SubmenuButton(
                    style: SubmenuButton.styleFrom(overlayColor: Colors.red),
                    menuChildren: [
                      MenuItemButton(
                        child: const Text('MenuItemButton'),
                        onPressed: () {},
                      ),
                    ],
                    child: const Text('SubmenuButton'),
                  ),
                ],
              ),
              SegmentedButton<Sizes>(
                style:
                    SegmentedButton.styleFrom(overlayColor: Colors.red),
                segments: const <ButtonSegment<Sizes>>[
                  ButtonSegment<Sizes>(
                      value: Sizes.extraSmall, label: Text('XS')),
                  ButtonSegment<Sizes>(value: Sizes.small, label: Text('S')),
                  ButtonSegment<Sizes>(value: Sizes.medium, label: Text('M')),
                  ButtonSegment<Sizes>(
                    value: Sizes.large,
                    label: Text('L'),
                  ),
                  ButtonSegment<Sizes>(
                      value: Sizes.extraLarge, label: Text('XL')),
                ],
                selected: const {Sizes.medium},
                onSelectionChanged: (Set<Sizes> newSelection) {},
                multiSelectionEnabled: true,
              ),
            ],
          ),
        ),
      ),
    );
  }
}
```

</details>

### Preview

ScreenRecording2024-04-12at15 25 58-ezgif com-video-to-gif-converter

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

**Comment by TahaTesser** [comment]
@HansMuller 
This PR adds the `overlayColor` prop you added in Buttons https://github.com/flutter/flutter/pull/142762

File: packages/flutter/lib/src/material/icon_button.dart [hunk 1]
```diff
@@ -629,20 +636,24 @@ class IconButton extends StatelessWidget {
...
+      ? null
+      : _IconButtonDefaultForeground(foregroundColor, disabledForegroundColor);
+    final MaterialStateProperty<Color?>? overlayColorProp = (foregroundColor == null &&
+      hoverColor == null && focusColor == null && highlightColor == null && overlayColor == null)
         ? null
-        : _IconButtonDefaultForeground(foregroundColor, disabledForegroundColor);
-    final MaterialStateProperty<Color?>? overlayColor = (foregroundColor == null && hoverColor == null && focusColor == null && highlightColor == null)
-        ? null
-        : _IconButtonDefaultOverlay(foregroundColor, focusColor, hoverColor, highlightColor);
+        : switch (overlayColor) {
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/icon_button.dart:647]: This switch statement is pretty slick. Are the parens in the else clause - `(_)` needed?

File: packages/flutter/lib/src/material/icon_button.dart [hunk 2]
```diff
@@ -1023,34 +1034,53 @@ class _IconButtonDefaultForeground extends MaterialStateProperty<Color?> {
...
   final Color? hoverColor;
   final Color? highlightColor;
+  final Color? overlayColor;
 
   @override
   Color? resolve(Set<MaterialState> states) {
     if (states.contains(MaterialState.selected)) {
       if (states.contains(MaterialState.pressed)) {
-        return highlightColor ?? foregroundColor?.withOpacity(0.1);
+        return highlightColor
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/icon_button.dart:1055]: This is fine as is. I suppose you could have also written (here and elsewhere):
```dart
return highlightColor ?? (overlayColor ?? foregroundColor)?.withOpacity(0.1);
```

**Review by HansMuller — APPROVED** [review]
Thanks for this update! LGTM

File: packages/flutter/lib/src/material/icon_button.dart (hunk 1 above)
Comment (TahaTesser) [reply packages/flutter/lib/src/material/icon_button.dart:647]: Good question, i checked other resources, They're not using parens. Fixed.

File: packages/flutter/lib/src/material/icon_button.dart (hunk 2 above)
Comment (TahaTesser) [reply packages/flutter/lib/src/material/icon_button.dart]: Perfect!

**Pull request merged**